### PR TITLE
upgrade API to target libuv 2

### PR DESCRIPTION
**Pull request by vtjnash**

### CODE REVIEW

File: src/init.c [hunk 1]
```diff
@@ -321,90 +317,138 @@ void *jl_winsock_handle;
...
 
-static void *init_stdio_handle(uv_file fd,int readable)
+#ifndef _OS_WINDOWS_
+#define UV_STREAM_READABLE 0x20   /* The stream is readable */
+#define UV_STREAM_WRITABLE 0x40   /* The stream is writable */
+#endif
+
+#ifdef _OS_WINDOWS_
+int uv_dup(uv_os_fd_t fd, uv_os_fd_t* dupfd) {
+  HANDLE current_process;
```
Comment (ararslan) [inline src/init.c]: Inconsistent indentation here, should be 4 space indents everywhere

**Comment by StefanKarpinski** [comment]
Really? Why now?

**Comment by ararslan** [comment]
This isn't really a user-facing change at all, is it?

**Comment by ararslan** [comment]
Never mind, there is a user facing change: it introduces a deprecation, `link_pipe` to `link_pipe!` (which should probably be `linkpipe!`).

**Comment by JeffBezanson** [comment]
Bump. It seems like a good idea to me to take this opportunity to upgrade libuv, rather than put it off even longer.

`link_pipe` is not exported. @vtjnash Can you say a bit more about the impact and prioritization of this? Could it be done in 1.x?

**Comment by vtjnash** [comment]
I was hoping to first rebase this PR against a more recent version of libuv master. There's a PR upstream that should get it up-to-date as-of October that seems like it might merge in the next few days to weeks.

**Comment by StefanKarpinski** [comment]
I don’t think we can afford to wait weeks for this so it seems like updating with this version would be better. We can always update again but this is currently on the 1.0 milestone because of behavioral changes.

**Comment by JeffBezanson** [comment]
What are the behavioral changes?

**Comment by StefanKarpinski** [comment]
I thought it was connected to https://github.com/JuliaLang/julia/issues/24717 but maybe it's not.

**Comment by vtjnash** [comment]
gtg?

**Comment by ViralBShah** [comment]
Is it possible to use a release version of libuv?

File: base/deprecated.jl [hunk 2]
```diff
@@ -1412,6 +1412,10 @@ end
 
 @deprecate print_with_color(color, args...; kwargs...) printstyled(args...; kwargs..., color=color)
 
+# PR 25062
+@deprecate(link_pipe(pipe; julia_only_read = true, julia_only_write = true),
+           link_pipe!(pipe, reader_supports_async = julia_only_read, writer_supports_async = julia_only_write))
```
Comment (JeffBezanson) [inline base/deprecated.jl]: Add `false` so it's not exported?

File: base/stream.jl [hunk 3]
```diff
@@ -168,7 +182,6 @@ mutable struct TTY <: LibuvStream
     lock::ReentrantLock
     throttle::Int
     @static if Sys.iswindows(); ispty::Bool; end
-    TTY() = TTY(Libc.malloc(_sizeof_uv_tty), StatusUninit)
```
Comment (JeffBezanson) [inline base/stream.jl]: Just curious: why does `PipeEndpoint` have a 0-arg constructor that calls malloc, but not `TTY`?

**Comment by JeffBezanson** [comment]
CI certainly seems to like this. Is there any chance this fixes the `spawn` test on mac?

**Comment by ararslan** [comment]
> Is it possible to use a release version of libuv?

At this point I don't believe so; we're still maintaining our own fork of libuv with Julia-specific changes and it's unclear whether they'll be upstreamed.

**Comment by vtjnash** [comment]
No, but it’s the prerequisite for https://github.com/JuliaLang/julia/pull/22041, which I hope will

**Comment by vtjnash** [comment]
> Is it possible to use a release version of libuv?

Not until they release a version of libuv v2.0, and merge our fixes.

**Comment by JeffBezanson** [comment]
I suppose this should be merged?

**Comment by JeffBezanson** [comment]
I'm going to merge this. We suspect it's fairly low risk, and the 0.7 timeframe is as good a time as any to try this out.

**Pull request merged**